<launch>
  <!-- General Parameters -->
  <arg name="PI" default="3.141592654"/>
  <arg name="PI_2" default="1.570796327"/>

  <!-- Custom -->
  <arg name="flag_gazebo" default="true"/>
  <arg name="flag_robot" default="true"/>
  <arg name="flag_moveit" default="false"/>
  <arg name="flag_conveyor" default="false"/>
  <arg name="flag_bin" default="false"/>
  <arg name="flag_pedsim" default="false"/>
  <arg name="flag_markers" default="false"/>
  <arg name="flag_map_server" default="false"/>
  <arg name="flag_rviz" default="false"/>

  <!-- Gazebo -->
  <arg name="paused" default="true"/>
  <arg name="use_sim_time" default="true"/>
  <arg name="gui" default="true"/>
  <arg name="headless" default="false"/>
  <arg name="debug" default="false"/>
  <arg name="verbose" default="false"/>
  <arg name="robot_namespace" default="/"/>

  <!-- 
    ur5
    j2n6s300
  -->
  <arg name="robot_arm" default="j2n6s300"/>

  <!-- Initial Robot Base Pose -->
  <arg name="robot0_init_pos_x" default="0.0"/>
  <arg name="robot0_init_pos_y" default="0.0"/>
  <arg name="robot0_init_pos_z" default="0.1"/>
  <arg name="robot0_init_roll"  default="0.0"/>
  <arg name="robot0_init_pitch" default="0.0"/>
  <arg name="robot0_init_yaw"   default="0.0"/>

  <!-- Launch the Gazebo world -->
  <group if="$(eval arg('flag_gazebo') == true)">
    <include file="$(find gazebo_ros)/launch/empty_world.launch">
      <arg name="world_name" value="$(find mobiman_simulation)/worlds/drl.world"/>
      <arg name="paused" value="$(arg paused)"/>
      <arg name="use_sim_time" value="$(arg use_sim_time)"/>
      <arg name="gui" value="$(arg gui)" />
      <arg name="headless" value="$(arg headless)"/>
      <arg name="debug" value="$(arg debug)" />
      <arg name="verbose" value="$(arg verbose)" />
    </include>

    <!-- Launch conveyor belt -->
    <group if="$(eval arg('flag_conveyor') == true)">
      <group ns="conveyor">
        <node name="spawn_conveyor_belt" 
              pkg="gazebo_ros" 
              type="spawn_model" 
              args="-file $(find mobiman_simulation)/urdf/conveyor_belt.urdf -urdf -model conveyor_belt -x 0.0 -y -2.5 -z 0.25" 
              respawn="false" 
              output="screen" />
      </group>
      <!-- <node name="spawn_packages" pkg="mobiman_simulation" type="spawn_packages.py" output="screen"/> -->
    </group>

    <!-- Launch bin -->
    <group if="$(eval arg('flag_bin') == true)">
      <group ns="bin_4_dropping_task">
        <node name="spawn_bin" 
              pkg="gazebo_ros" 
              type="spawn_model" 
              args="-file $(find mobiman_simulation)/models/bin_4_dropping_task/model.sdf -sdf -model bin_4_dropping_task 
                -x 4.0 -y 0.0 -z 0.1
                -R 0.0 -P 0.0 -Y $(arg PI_2)" 
              respawn="false" 
              output="screen" />
      </group>
    </group>
  </group>

  <!-- Launch robot -->
  <group if="$(eval arg('flag_robot') == true)">
    <group ns="$(arg robot_namespace)">
      <!-- Load the URDF into the ROS Parameter Server -->
      <group if="$(eval arg('robot_arm') == 'ur5')">
        <param name="robot_description" command="$(find xacro)/xacro '$(find mobiman_simulation)/urdf/jackal_ur5.urdf.xacro'" />
      </group>
      
      <group if="$(eval arg('robot_arm') == 'j2n6s300')">
        <param name="robot_description" command="$(find xacro)/xacro '$(find mobiman_simulation)/urdf/jackal_jaco.xacro'" />
      </group>

      <group if="$(eval arg('flag_moveit') == true)">
        <include file="$(find mobiman_simulation)/launch/utilities/moveit_planner.launch">
        </include>
      </group>

      <!-- Run a python script to the send a service call to gazebo_ros to spawn a URDF robot -->
      <group if="$(eval arg('robot_arm') == 'ur5')">
        <arg name="name_arm_joint_1" default="ur5_shoulder_pan_joint"/>
        <arg name="name_arm_joint_2" default="ur5_shoulder_lift_joint"/>
        <arg name="name_arm_joint_3" default="ur5_elbow_joint"/>
        <arg name="name_arm_joint_4" default="ur5_wrist_1_joint"/>
        <arg name="name_arm_joint_5" default="ur5_wrist_2_joint"/>
        <arg name="name_arm_joint_6" default="ur5_wrist_3_joint"/>

        <arg name="init_arm_joint_pos_1" value="0.0"/>
        <arg name="init_arm_joint_pos_2" value="2.9"/>
        <arg name="init_arm_joint_pos_3" value="1.3"/>
        <arg name="init_arm_joint_pos_4" value="4.2"/>
        <arg name="init_arm_joint_pos_5" value="1.4"/>
        <arg name="init_arm_joint_pos_6" value="0.0"/>

        <!--arg name="init_arm_joint_pos_1" default="0.0"/>
        <arg name="init_arm_joint_pos_2" default="0.0"/>
        <arg name="init_arm_joint_pos_3" default="0.0"/>
        <arg name="init_arm_joint_pos_4" default="0.0"/>
        <arg name="init_arm_joint_pos_5" default="0.0"/>
        <arg name="init_arm_joint_pos_6" default="0.0"/-->

        <node name="urdf_spawner" 
              pkg="gazebo_ros" 
              type="spawn_model" 
              respawn="false" 
              output="screen"
              args="-urdf 
                    -model mobiman 
                    -param robot_description 
                    -robot_namespace $(arg robot_namespace)
                    -x $(arg robot0_init_pos_x)
                    -y $(arg robot0_init_pos_y)
                    -z $(arg robot0_init_pos_z)
                    -R $(arg robot0_init_roll)
                    -P $(arg robot0_init_pitch)  
                    -Y $(arg robot0_init_yaw)
                    -J $(arg name_arm_joint_1) $(arg init_arm_joint_pos_1)
                    -J $(arg name_arm_joint_2) $(arg init_arm_joint_pos_2)
                    -J $(arg name_arm_joint_3) $(arg init_arm_joint_pos_3)
                    -J $(arg name_arm_joint_4) $(arg init_arm_joint_pos_4)
                    -J $(arg name_arm_joint_5) $(arg init_arm_joint_pos_5)
                    -J $(arg name_arm_joint_6) $(arg init_arm_joint_pos_6)
                    -unpause"/>
      </group>
      
      <group if="$(eval arg('robot_arm') == 'j2n6s300')">

        <arg name="name_arm_joint_1" value="j2n6s300_joint_1"/>
        <arg name="name_arm_joint_2" value="j2n6s300_joint_2"/>
        <arg name="name_arm_joint_3" value="j2n6s300_joint_3"/>
        <arg name="name_arm_joint_4" value="j2n6s300_joint_4"/>
        <arg name="name_arm_joint_5" value="j2n6s300_joint_5"/>
        <arg name="name_arm_joint_6" value="j2n6s300_joint_6"/>

        <arg name="init_arm_joint_pos_1" value="0.0"/>
        <arg name="init_arm_joint_pos_2" value="2.9"/>
        <arg name="init_arm_joint_pos_3" value="1.3"/>
        <arg name="init_arm_joint_pos_4" value="4.2"/>
        <arg name="init_arm_joint_pos_5" value="1.4"/>
        <arg name="init_arm_joint_pos_6" value="0.0"/>

        <node name="urdf_spawner" 
              pkg="gazebo_ros" 
              type="spawn_model" 
              respawn="false" 
              output="screen"
              args="-urdf 
                    -model mobiman 
                    -param robot_description 
                    -robot_namespace $(arg robot_namespace)
                    -x $(arg robot0_init_pos_x)
                    -y $(arg robot0_init_pos_y)
                    -z $(arg robot0_init_pos_z)
                    -R $(arg robot0_init_roll)
                    -P $(arg robot0_init_pitch)  
                    -Y $(arg robot0_init_yaw)
                    -J $(arg name_arm_joint_1) $(arg init_arm_joint_pos_1)
                    -J $(arg name_arm_joint_2) $(arg init_arm_joint_pos_2)
                    -J $(arg name_arm_joint_3) $(arg init_arm_joint_pos_3)
                    -J $(arg name_arm_joint_4) $(arg init_arm_joint_pos_4)
                    -J $(arg name_arm_joint_5) $(arg init_arm_joint_pos_5)
                    -J $(arg name_arm_joint_6) $(arg init_arm_joint_pos_6)"/>
        </group>

      <!--node name="publish_ground_truth_odom" pkg="mobiman_simulation" type="publish_ground_truth_odom.py" output="screen" /-->
      <node name="world_frame_server" pkg="mobiman_simulation" type="world_frame_server" output="screen" />

      <!-- node -->
      <node name="reset_mobiman" pkg="mobiman_simulation" type="reset_mobiman.py" output="screen" />
      
      <!-- Load Robot Semantic -->
      <!--param name="robot_description_semantic" textfile="$(find mobiman_wb_moveit_config)/config/mobiman.srdf" /-->

      <!-- Run the sensor processes -->
      <!--include file="$(find mobiman_simulation)/launch/utilities/camera_process.launch" /-->

      <!-- Run robot control on the gazebo master -->
      <group if="$(eval arg('robot_arm') == 'ur5')">
        <include file="$(find mobiman_control)/launch/control_jackal_ur5.launch" />
      </group>

      <group if="$(eval arg('robot_arm') == 'j2n6s300')">
        <include file="$(find mobiman_control)/launch/control_jackal_jaco.launch" />

        <!--node name="command_robot_home_pose" pkg="mobiman_simulation" type="move_jaco.py"
          respawn="false" output="screen" args="$(arg robot_arm)">       
        </node-->
      </group>
    </group>
  </group>

  <!-- Run Pedsim -->
  <group if="$(eval arg('flag_gazebo') == true)">
    <group if="$(eval arg('flag_pedsim') == true)">
      <node pkg="mobiman_simulation" type="spawn_pedsim_agents.py" name="spawn_pedsim_agents" output="screen" />
      <include file="$(find mobiman_simulation)/launch/utilities/pedsim.launch" />
    </group>
  </group>

  <!-- Spawn marker on the robot -->
  <!--node if="$(arg flag_markers)" name="spawn_markers" pkg="mobiman_simulation" type="spawn_markers" output="screen" /-->

  <!-- Run Map Server -->
  <!--group if="$(eval arg('flag_gazebo') == true)">
    <group if="$(eval arg('flag_map_server') == true)">
      <include file="$(find mobiman_simulation)/launch/utilities/map_server.launch" />
    </group>
  </group-->

  <!-- Run Rviz visualization -->
  <!--group if="$(eval arg('flag_rviz') == true)">
    <include file="$(find mobiman_simulation)/launch/examples/rviz_ocs2_gazebo.launch" />
  </group-->
</launch>
